### ISSUE TITLE
[WebGPU] GroupQueryAttention in-place alias breaks growing KV caches; Conv activation fusion outruns the ep-webgpu plug-in

### ISSUE THREAD

**Issue by ykhrustalev** [microsoft/onnxruntime#32716]
Two independent WebGPU EP failures found while bringing a speech model up on the plug-in EP. They
are unrelated to each other; happy to split them into separate issues if you'd prefer — filing
together because they were found in one investigation and share an environment.

Both reproduce on **Metal (macOS/arm64)** and **Vulkan/lavapipe (Linux/x64)**, in pure
onnxruntime with no other runtime in the process, with minimal synthetic models that need no
downloads. In every WebGPU run below `session.get_providers()` was asserted to contain
`WebGpuExecutionProvider`, so none of these are a silent CPU fallback.

| Machine | onnxruntime | onnxruntime-ep-webgpu | Backend |
|---|---|---|---|
| Ubuntu 22.04, x86_64 | 1.31.0.dev20260915003 | 0.3.0 | Dawn/Vulkan on lavapipe |
| macOS 26.6.2, M3 Ultra | 1.31.0.dev20260920001 | 0.3.0 | Metal |

EP selection used throughout:

```python
import onnxruntime as ort, onnxruntime_ep_webgpu as w
ort.register_execution_provider_library(w.get_ep_name(), w.get_library_path())
so = ort.SessionOptions()
so.add_provider_for_devices([d for d in ort.get_ep_devices() if d.ep_name == w.get_ep_name()][:1], {})
```

(The `[:1]` is needed on the Linux box: it enumerates two WebGPU devices, passing both raises
"WebGPU EP factory currently only supports one device at a time", and the session then falls back
to CPU — which is easy to mistake for success.)

---

# 1. GroupQueryAttention declares an unconditional in-place alias, so any growing KV cache aborts

`CreateGroupQueryAttentionKernelInfo` in `contrib_ops/webgpu/bert/group_query_attention.cc` declares

```cpp
      .MayInplace(3, 1)   // past_key   -> present_key
      .MayInplace(4, 2)   // past_value -> present_value
```

unconditionally. That alias is only valid when past and present share a pre-allocated buffer. When
the planner is free to honour it — i.e. the past KV tensor is a planner-owned intermediate rather
than a user-supplied graph input — and the kernel then needs `present_seq_len > past_seq_len`, the
run aborts:

```
Non-zero status code returned while running GroupQueryAttention node. Name:'/layers.0/gqa'
Status Message: core/framework/op_kernel.cc:83 OpKernelContext::OutputMLValue(int, const TensorShape&)
[ONNXRuntimeError] : 1 : FAIL : Shape mismatch attempting to re-use buffer. {1,8,0,32} != {1,8,1,32}.
```

The CPU GQA kernel declares no alias and runs the same graph and feeds correctly.

### The rule, from a grid over (past_buffer_len, seqlens_k, total_seq_len)

| past buffer | seqlens_k | total_seq_len | WebGPU | CPU |
|---|---|---|---|---|
| 0 | 0 | 1 | **FAIL** `{1,2,0,32} != {1,2,1,32}` | OK |
| 1 | 1 | 2 | **FAIL** `{1,2,1,32} != {1,2,2,32}` | OK |
| 2 | 2 | 3 | **FAIL** `{1,2,2,32} != {1,2,3,32}` | OK |
| 1 | 0 | 1 | OK | OK |
| 4 | 3 | 4 | OK | OK |
| 4 | 0 | 4 | OK | OK |

It is not specific to the first (empty-past) step: it fails on every step where the present must
grow, so a grow-by-concat KV cache fails at *every* decode step. Where WebGPU does run, its output
matches CPU (`logits[0,:3] = -0.89720, 1.68429, -0.99122` on both).

It is also not "GQA on WebGPU" in general: a single-GQA graph whose `past_key`/`past_value` are
**graph inputs** runs fine with a growing present (confirmed by profiling that the node was assigned
to `WebGpuExecutionProvider`, with and without `do_rotary`). The trigger is the past KV being a
planner-owned intermediate — in the model where we hit it, `Gather(past_keys, layer_idx)`.

Unaffected by `ORT_DISABLE_ALL`, `ORT_ENABLE_BASIC`, `enable_mem_pattern=False`,
`enable_cpu_mem_arena=False`, `enable_mem_reuse=False` and `ORT_SEQUENTIAL` — consistent with a
plan-time decision.

### Repro — 5 nodes, no model download

```python
import sys, numpy as np, onnx
from onnx import helper as h, TensorProto as T, numpy_helper as nh
import onnxruntime as ort

B, S, NH, KVH, D = 1, 1, 4, 2, 32
hid, khid, L = NH * D, KVH * D, 2
ep, mp = sys.argv[1], "/tmp/gqa_min.onnx"
nodes = [
    h.make_node("Gather", ["past_keys", "idx"], ["past_key"], axis=0),
    h.make_node("Gather", ["past_values", "idx"], ["past_value"], axis=0),
    h.make_node("GroupQueryAttention",
                ["q", "k", "v", "past_key", "past_value", "seqlens_k", "total_seq_len"],
                ["out", "present_key", "present_value"], domain="com.microsoft",
                num_heads=NH, kv_num_heads=KVH),
    h.make_node("Unsqueeze", ["present_key", "ax"], ["new_keys"]),
    h.make_node("Unsqueeze", ["present_value", "ax"], ["new_values"]),
]
g = h.make_graph(nodes, "g",
    [h.make_tensor_value_info("q", T.FLOAT, [B, S, hid]),
     h.make_tensor_value_info("k", T.FLOAT, [B, S, khid]),
     h.make_tensor_value_info("v", T.FLOAT, [B, S, khid]),
     h.make_tensor_value_info("past_keys", T.FLOAT, [L, B, KVH, "p", D]),
     h.make_tensor_value_info("past_values", T.FLOAT, [L, B, KVH, "p", D]),
     h.make_tensor_value_info("seqlens_k", T.INT32, [B]),
     h.make_tensor_value_info("total_seq_len", T.INT32, [])],
    [h.make_tensor_value_info("out", T.FLOAT, [B, S, hid]),
     h.make_tensor_value_info("new_keys", T.FLOAT, [1, B, KVH, "n", D]),
     h.make_tensor_value_info("new_values", T.FLOAT, [1, B, KVH, "n", D])],
    initializer=[nh.from_array(np.array(0, np.int64), "idx"),
                 nh.from_array(np.array([0], np.int64), "ax")])
m = h.make_model(g, opset_imports=[h.make_opsetid("", 21), h.make_opsetid("com.microsoft", 1)])
m.ir_version = 10
onnx.save(m, mp)

so = ort.SessionOptions(); so.log_severity_level = 4
if ep == "webgpu":
    import onnxruntime_ep_webgpu as w
    ort.register_execution_provider_library(w.get_ep_name(), w.get_library_path())
    so.add_provider_for_devices([d for d in ort.get_ep_devices() if d.ep_name == w.get_ep_name()][:1], {})
    s = ort.InferenceSession(mp, so)
else:
    s = ort.InferenceSession(mp, so, providers=["CPUExecutionProvider"])
print("PROVIDERS:", s.get_providers())

rng = np.random.default_rng(0)
for past, sk, tot in [(0, 0, 1), (1, 1, 2), (1, 0, 1), (4, 3, 4), (2, 2, 3), (4, 0, 4)]:
    feeds = {"q": rng.standard_normal((B, S, hid), dtype=np.float32),
             "k": rng.standard_normal((B, S, khid), dtype=np.float32),
             "v": rng.standard_normal((B, S, khid), dtype=np.float32),
             "past_keys": rng.standard_normal((L, B, KVH, past, D), dtype=np.float32),
             "past_values": rng.standard_normal((L, B, KVH, past, D), dtype=np.float32),
             "seqlens_k": np.array([sk], np.int32), "total_seq_len": np.array(tot, np.int32)}
    try:
        out = s.run(None, feeds)
        print(f"  OK   past={past} sk={sk} tot={tot} new_keys={out[1].shape}")
    except Exception as e:
        print(f"  FAIL past={past} sk={sk} tot={tot} {str(e)[-120:]}")
```

Identical output on both machines: rows 1, 2 and 5 fail on WebGPU and pass on CPU; the rest pass on
both.

### Expected

Either make the in-place conditional on past/present actually sharing a buffer, or drop the
`MayInplace` declarations so the planner allocates a correctly sized present. Workaround for users:
pre-allocate the KV buffer to the maximum total length and run in share-buffer mode.

---

# 2. Conv activation fusion outruns the `onnxruntime-ep-webgpu` plug-in, and an unparseable activation kills session creation

Creating a session for a model containing `Conv → SiLU` fails at `ORT_ENABLE_EXTENDED` and above:

```
[ONNXRuntimeError] : 11 : EP_FAIL : core/providers/webgpu/nn/conv.h:21
onnxruntime::webgpu::Conv<is_channels_last, is_fused>::Conv(const OpKernelInfo&)
[with bool is_channels_last = true; bool is_fused = true]
GetFusedActivationAttr(info, activation_).IsOK() was false.
```

### Optimisation level decides it

| Level | Linux | macOS |
|---|---|---|
| `ORT_DISABLE_ALL` | session created | session created |
| `ORT_ENABLE_BASIC` | **created, runs on WebGPU, correct output** | **created, runs on WebGPU, correct output** |
| `ORT_ENABLE_EXTENDED` | EP_FAIL | EP_FAIL |
| `ORT_ENABLE_ALL` (default) | EP_FAIL | EP_FAIL |

At BASIC on a real model (a 56-Conv speech encoder) every Conv node is assigned to
`WebGpuExecutionProvider` — 1373 WebGPU / 628 CPU nodes — and the output is correct. **The WebGPU
Conv kernel is fine; the fused-activation attribute it is handed is the problem.**

### Which activations

Probing one activation at a time with a 3-node model:

- **Pass** at every level: bare Conv, `Relu`, `Sigmoid`, `Tanh`, `LeakyRelu`, `HardSigmoid`
- **Fail** at EXTENDED and above: `QuickGelu`, `Gelu` (onnx-20 and com.microsoft), `FastGelu`,
  `HardSwish`, `Elu`, `Softplus`, `ThresholdedRelu`, `Erf`

That failing set is exactly the list added by
[#32117 "[WebGPU] Fuse eight more activations into Conv"](https://github.com/microsoft/onnxruntime/pull/32117),
merged **2026-08-30**, which changed both `core/optimizer/conv_activation_fusion.cc` (ships in the
`onnxruntime` core wheel) and `core/providers/webgpu/nn/fuse_utils.{cc,h}` (ships in the
`onnxruntime-ep-webgpu` plug-in binary).

`onnxruntime-ep-webgpu` **0.3.0 was published to PyPI on 2026-08-24**, six days before that PR
merged, and is still the latest release. So a current core wheel emits `activation="QuickGelu"` and
the shipped plug-in's `GetFusedActivationAttr` rejects it. `strings` on the plug-in `.so`/`.dylib`
finds no `Erf` or `Elu` on either platform, consistent with it predating the change.

So the first ask is simply: **publish a plug-in release built from a commit at or after #32117.**

### The robustness bug underneath it

`conv.cc` registers `Conv` in `kMSInternalNHWCDomain` as `Conv<true, true>` — the plain NHWC Conv
uses the *fused* template — and the constructor `ORT_ENFORCE`s on `GetFusedActivationAttr`. An
activation string the kernel cannot parse therefore **aborts session creation** rather than the EP
declining the node and letting it fall to CPU. Any future core/plug-in skew, or any third-party
fusion emitting an unknown activation, becomes a hard load failure instead of a slower model. This
is worth fixing independently of the version skew.

Also: `session.disable_specified_optimizers` does **not** avoid it. With `ConvActivationFusion`
listed, verbose logs still show `Applying graph transformer ConvActivationFusion on step 2 ...
modified: 1` and the session still fails — the post-layout-transform re-run appears to ignore the
disable list. (`ConvAddActivationFusion`, `NhwcTransformer` and `QuickGeluFusion` made no difference
either.) That looks like a third, smaller bug.

### Repro — 3 nodes, no model download

```python
import numpy as np, onnx
from onnx import helper, TensorProto
import onnxruntime as ort, onnxruntime_ep_webgpu as w

C = 8
W = np.random.randn(C, C, 3).astype(np.float32)
B = np.random.randn(C).astype(np.float32)
nodes = [
    helper.make_node("Conv", ["X", "W", "Bi"], ["c"], kernel_shape=[3],
                     pads=[1, 1], strides=[1], dilations=[1], group=1),
    helper.make_node("Sigmoid", ["c"], ["s"]),
    helper.make_node("Mul", ["c", "s"], ["y"]),          # SiLU
]
g = helper.make_graph(nodes, "g",
    [helper.make_tensor_value_info("X", TensorProto.FLOAT, [1, C, 16])],
    [helper.make_tensor_value_info("y", TensorProto.FLOAT, None)],
    initializer=[helper.make_tensor("W", TensorProto.FLOAT, W.shape, W.flatten()),
                 helper.make_tensor("Bi", TensorProto.FLOAT, B.shape, B.flatten())])
m = helper.make_model(g, opset_imports=[helper.make_opsetid("", 17)]); m.ir_version = 8
onnx.save(m, "/tmp/conv_silu.onnx")

ort.register_execution_provider_library(w.get_ep_name(), w.get_library_path())
so = ort.SessionOptions(); so.log_severity_level = 3
so.graph_optimization_level = ort.GraphOptimizationLevel.ORT_ENABLE_EXTENDED   # BASIC -> works
so.add_provider_for_devices([d for d in ort.get_ep_devices() if d.ep_name == w.get_ep_name()][:1], {})
s = ort.InferenceSession("/tmp/conv_silu.onnx", so)      # EP_FAIL here
print(s.get_providers())
```

### Expected

1. A plug-in release built at or after #32117, so core and plug-in agree on the activation set.
2. The NHWC `Conv` kernel should decline a node whose activation it cannot parse, rather than
   `ORT_ENFORCE`-ing during session initialisation.
3. `session.disable_specified_optimizers` should be honoured by the post-layout-transform
   transformer run.
